api/include/opentelemetry/trace/trace_state.h

```diff
@@ -45,51 +45,154 @@ class TraceState
   class Entry
   {
   public:
-    Entry() noexcept = default;
-
-    // Copy constructor.
-    Entry(const Entry &copy);
-
-    // Creates an Entry for a given key-value pair.
-    Entry(nostd::string_view key, nostd::string_view value) noexcept;
-
-    nostd::string_view GetKey();
-    nostd::string_view GetValue();
+    Entry() : key_(new char[kKeyMaxSize]), value_(new char[kValueMaxSize]){};
+
+    // Copy constructor
+    Entry(Entry &copy)
+    {
+      key_   = StringToPointer(copy.key_.get(), kKeyMaxSize);
+      value_ = StringToPointer(copy.value_.get(), kValueMaxSize);
+    }
+
+    // Assignment operator
+    Entry &operator=(Entry &other)
+    {
+      key_   = StringToPointer(other.key_.get(), kKeyMaxSize);
+      value_ = StringToPointer(other.value_.get(), kValueMaxSize);
+      return *this;
+    }
+
+    // Create an Entry for a given key-value pair.
+    Entry(nostd::string_view key, nostd::string_view value) noexcept
+    {
+      key_   = StringToPointer(key.data(), kKeyMaxSize);
+      value_ = StringToPointer(value.data(), kValueMaxSize);
+    }
+
+    // Get the key associated with this entry.
+    nostd::string_view GetKey() { return nostd::string_view(key_.get()); }
+
+    // Get the value associated with this entry.
+    nostd::string_view GetValue() { return nostd::string_view(value_.get()); }
+
+    // Set the value for this entry. This overrides the previous value.
+    void SetValue(nostd::string_view value)
+    {
+      value_ = StringToPointer(value.data(), kValueMaxSize);
+    }
 
   private:
     // Store key and value as raw char pointers to avoid using std::string.
-    nostd::unique_ptr<const char[]> key_;
-    nostd::unique_ptr<const char[]> value_;
+    std::unique_ptr<const char[]> key_;
+    std::unique_ptr<const char[]> value_;
+
+    // Copy string into a buffer and return a unique_ptr to the buffer.
+    // This is a workaround for the fact that strcpy doesn't accept a const char* destination.
+    nostd::unique_ptr<const char[]> StringToPointer(const char *str, const int kMaxSize)
+    {
+      nostd::unique_ptr<char[]> temp(new char[kMaxSize]);
+      strcpy(temp.get(), str);
+      return nostd::unique_ptr<const char[]>(temp.release());
+    }
```

Comment:
Could we let this function take a `nostd::string_view` as argument? This encapsulates the char pointer and its size, so you'd not need to pass those arguments separately. Furthermore, you can allocate only what's necessary.
```c++
    nostd::unique_ptr<const char[]> StringToPointer(nostd::string_view str)
    {
      char temp[] = new char[str.size()];
      strcpy(temp, str.data());
      return nostd::unique_ptr<const char[]>(temp);
    }
```

If we want to go with always allocating `kKeyMaxSize` and `kValueMaxSize`, we could just define the class members with those fixed sizes:
```c++
const char key_[kKeyMaxSize];
const char value_[kValueMaxSize];
```

Reply:
Yes, I thought about making the function take a `nostd::string_view`, but then when I call the function from the copy constructor or assignment operator, I would have to create `string_view` instances for `key_` and `value_`. Then, in `StringToPointer`, all we really need are the raw `char*`. However, I agree that only allocating the memory that is necessary would be better.

What are your thoughts on always allocating `kKeyMaxSize` and `kValueMaxSize`? I thought this might be simpler and more robust, but the way the implementation turned out I don't think it's really necessary.

Reply:
Perhaps it would make sense to store `string_view` instances of `key_` and `value_` in `Entry`?

Reply:
I've updated `StringToPointer()` to take a `string_view`.

Reply:
We shouldn't store `string_view` instances, as those are non-owning references, they don't allow us to control the lifetime of the data.